in-place slice matrix multiplications

ffreyer opened the pull request:
This pr changes every function in `DQMC/slice_matrices.jl` to be in-place. (As in `DQMC.jl`) It also adds tests for each function in `DQMC/slice_matrices.jl`

For some reason `multiply_slice_matrix_right!` and `multiply_slice_matrix_inv_right!` for `::DQMC_CBTrue` are (much) slower than on master atm. They use more than double the allocations of their `left!` counterparts. See benchmarks: 
https://gist.github.com/ffreyer/58e3e3d9b7f42694d0917dac9860c980

TODO:
- [x] Fix performance regression for `multiply_slice_matrix_right!(::DQMC_CBTrue, ...)` and `multiply_slice_matrix_inv_right!(::DQMC_CBTrue, ...)`

ffreyer commented:
The reason for the different performance between `multiply_slice_matrix_left!` and `multiply_slice_matrix_right!` are the `mul!` calls. `mul!(::Matrix, ::Matrix, ::SparseMatrixCSC)` has more allocations and longer runtime than `mul!(::Matrix, ::SparseMatrixCSC, ::Matrix)` (which has 0 allocations).
The former ends up calling (slow)
https://github.com/JuliaLang/julia/blob/8ebe5643ca009af913c2a4b7b8707e2c0c7c69df/stdlib/LinearAlgebra/src/matmul.jl#L692.
while the latter calls (fast)
https://github.com/JuliaLang/julia/blob/8ebe5643ca009af913c2a4b7b8707e2c0c7c69df/stdlib/SparseArrays/src/linalg.jl#L34

crstnbr commented:
Note that this is https://discourse.julialang.org/t/asymmetric-speed-of-in-place-sparse-dense-matrix-product/10256/3. Can be temporarily fixed by
```julia
import LinearAlgebra.mul!
function mul!(C::StridedMatrix, X::StridedMatrix, A::SparseMatrixCSC)
    mX, nX = size(X)
    nX == A.m || throw(DimensionMismatch())
    fill!(C, zero(eltype(C)))
    rowval = A.rowval
    nzval = A.nzval
    @inbounds for  col = 1:A.n, k=A.colptr[col]:(A.colptr[col+1]-1)
        ki=rowval[k]
        kv=nzval[k]
        for multivec_row=1:mX
            C[multivec_row, col] += X[multivec_row, ki] * kv
        end
    end
    C
end
```

ffreyer commented:
I'll add it to the pr with some tests. I'm thinking of adding a test like 
```julia
@test (@which mul!(X, M, S)).module == LinearAlgebra
```
so we get reminded when the julia pr passes

ffreyer commented:
I added a remidner test that checks the type signature of `mul!`. I also added some checks for `right_mul!` and some notes on why it's there.

With that, the pr is now a performance boost across the board. In fact `multiply_slice_matrix_(inv)_right!` is now about twice as fast as the `left` variants.

ffreyer commented:
I think this is ready to be merged

carstenbauer commented on src/helpers.jl:
```diff
@@ -65,3 +65,24 @@ function compare(A::AbstractArray{T}, B::AbstractArray{S}) where T<:Number where
+# vurrenlty julia/sparearrays does not implement this function (type signature)
+# once it does this can be removed/depracted in favor of mul!
+# see also: test/slice_matrices.jl
+function right_mul!(C::StridedMatrix, X::StridedMatrix, A::SparseMatrixCSC)
```
Why not overload `LinearAlgebra.mul!`? We then only have to remove the custom definition here once it lands in Julia.

carstenbauer commented on test/slice_matrices.jl:
```diff
@@ -0,0 +1,134 @@
+    M = rand(4, 4)
+    S = sparse(rand(4, 4))
+    X = similar(M)
+    mul_dense_sparse_is_missing = (@which mul!(X, M, S)).sig == Tuple{typeof(mul!),AbstractArray{T,2} where T,Union{AbstractArray{T,1}, AbstractArray{T,2}} where T,Union{AbstractArray{T,1}, AbstractArray{T,2}} where T}
```
I like the idea of having a notification once the specialization lands in Julia. However, I wouldn't do it at the expense of inventing a new name `right_mul!` above. Also, the tests probably should fail because of this. The code works regardless of who defines the specialization (even if no ones defines it). Maybe it's better to throw a `@warn` in helpers.jl upon the definiton of the specialization in case it already exists in `LinearAlgebra`?

carstenbauer replied:
Instead of creating fake `M, S, X` you can just call `which(mul!, (Matrix, Matrix, SparseMatrixCSC))`.

carstenbauer replied:
Checking types, like you do, is one option. Another rather solid check could be
`occursin("SparseArrays", string(which(mul!, (Matrix, Matrix, SparseMatrixCSC)).file)) == false`

ffreyer replied:
I added such a warning now and switched to `mul!` and `occursin(...)`.

carstenbauer commented on test/slice_matrices.jl:
```diff
@@ -0,0 +1,134 @@
+        MonteCarlo.interaction_matrix_exp!(dqmc, m, eV, dqmc.conf, slice, 1.)
+
+        # MonteCarlo.slice_matrix
+        @test maximum(abs.(
```
Why not simply `maximum(abs, A-B)` (avoid broadcasting and allocations)?
